homeassistant/components/asuswrt/config_flow.py

```diff
@@ -0,0 +1,175 @@
+                        {"ssh": "SSH", "telnet": "Telnet"}
+                    ),
+                    vol.Required(CONF_PORT, default=DEFAULT_SSH_PORT): cv.port,
+                    vol.Required(CONF_MODE, default="router"): vol.In(
+                        {"router": "Router", "ap": "Access Point"}
+                    ),
+                    vol.Required(CONF_REQUIRE_IP, default=True): bool,
+                    vol.Required(CONF_INTERFACE, default=DEFAULT_INTERFACE): str,
+                    vol.Required(CONF_DNSMASQ, default=DEFAULT_DNSMASQ): str,
+                }
+            ),
+            errors=errors or {},
+        )
+
+    async def _async_check_connection(self, user_input):
+        """Attempt to connect the AsusWrt router."""
+
+        errors = {}
+
+        api = get_api(user_input)
+        try:
+            await api.connection.async_connect()
+            if api.is_connected:
+                if hasattr(api.connection, "disconnect"):
+                    await api.connection.disconnect()
+
+                return self.async_create_entry(
+                    title=self._name,
+                    data=user_input,
+                )
+
+            _LOGGER.error("Error connecting to the AsusWrt router at %s", self._host)
+            errors["base"] = "cannot_connect"
+
+        except OSError:
+            _LOGGER.error("Error connecting to the AsusWrt router at %s", self._host)
+            errors["base"] = "cannot_connect"
+
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception(
+                "Unknown error connecting with AsusWrt router at %s", self._host
+            )
+            errors["base"] = "unknown"
+
+        return self._show_setup_form(user_input, errors)
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initiated by the user."""
+        if self._async_current_entries():
+            return self.async_abort(reason="single_instance_allowed")
+
+        errors = {}
+
+        if user_input is None:
+            return self._show_setup_form(user_input, errors)
+
+        host = user_input[CONF_HOST]
+        pwd = user_input.get(CONF_PASSWORD)
+        ssh = user_input.get(CONF_SSH_KEY)
+
+        if not (pwd or ssh):
+            errors["base"] = "pwd_or_ssh"
+        elif ssh:
+            if pwd:
+                errors["base"] = "pwd_and_ssh"
+            else:
+                isfile = await self.hass.async_add_executor_job(_is_file, ssh)
+                if not isfile:
+                    errors["base"] = "ssh_not_file"
+
+        if not errors:
+            ip_address = await self.hass.async_add_executor_job(_get_ip, host)
+            if not ip_address:
+                errors["base"] = "invalid_host"
+
+        if errors:
+            return self._show_setup_form(user_input, errors)
+
+        self._host = host
+        self._name = user_input.get(CONF_NAME, host)
```

Comment:
We don't need to ask for name, I think.

Reply:
The idea was to have a more clear name, because host name should be an IP, but that's ok.

Related to the single config entry, this is not fully true. Let me say that I should be able to manage multiple config entry but I need to implement the part related to the `scanner_entity`, that we choose to introduce as second PR.
But the only unique_id that I can use is the IP Address and as you stated in next comment it is not allowed.

Reply:
Ok.

We can use the config entry id as unique id I think. It's not limited to a single entry, I was thinking of DOMAIN as unique id in that case. I'm not sure why I wrote that.

Reply:
Sorry, little be confused here. How can I get the entry id to set it as unique id if the entry is still not created?

Reply:
I think that I cannot manage multiple entry if I don't have a real unique ID for the device (e.g. MAC ADDRESS). The HostName  and IPAddress can change, at the end it should be possible to have an entry related to an IP while it is using another one because IP is changed. Is very rare case that a router change the IP, but it could be. Same consideration can be done for hostname.
I could use a package like [this](https://pypi.org/project/getmac/0.6.0/) to determinate mac address during setup but I don't now if is allowed and make sense.

Reply:
I'm asking about the entity unique_id not the config entry unique_id. The entities will be created after the config entry so we'llkknow the config entry id.

Reply:
Yes, that's fine, probably this comment is more related to other comment. I report there